### PR TITLE
[java] Fix #5253: Support Boolean wrapper class for BooleanGetMethodName rule

### LINKED ISSUE #5253: [java] BooleanGetMethodName: False-negatives with `Boolean` wrapper
**Is your feature request related to a problem? Please describe.**
Methods using `Boolean` return declaration are not managed by the `BooleanGetMethodName` rule.

**Describe the solution you'd like**
Update the rule to consider the `Boolean` wrapper too.

**Describe alternatives you've considered**
If this is considered as a BC break, I guess a property enabling the detection for the `Boolean` type could work.

**Additional context**
```java
Boolean getFoo() {
    return true;
}
```

https://github.com/pmd/pmd/blob/a0818d5ab2669dfd7f6bf8ca9e0ec8d8e0148b5d/pmd-java/src/main/resources/category/java/codestyle.xml#L187-L192

### PR DESCRIPTION
**Pull request by Aryant-Tripathi**
- Added Support Boolean wrapper class for BooleanGetMethodName rule (#5253)

**Changes** 
- Updated the XPath expression in the rule definition to also check for methods returning the Boolean wrapper class, ensuring that methods like getBooleanValue() are flagged the same way as methods returning boolean.
- This allows for greater consistency in applying the rule to both boolean and Boolean return types.

- Fixes #5253

### CODE REVIEW

**Review by punkratz312 — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Comment by jsotuyod** [comment]
@Aryant-Tripathi thanks for the PR.

This however will require tests. Test cases for rules are added as XML snippets, in this case on the [`BooleanGetMethodName.xml`](https://github.com/pmd/pmd/blob/main/pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml) file.

Also, if I recall correctly, there is no `@Type` attribute in `ReferenceType`, nor would that check work (think `Boolean` vs `java.lang.Boolean`). You will probably want to use [`pmd-java:typeIs("java.lang.Boolean")`](https://docs.pmd-code.org/latest/pmd_userdocs_extending_writing_xpath_rules.html#pmd-java-typeIs) instead to ensure proper coverage…

**Comment by oowekyala** [comment]
Also ReferenceType is not an xpath node in pmd 7. It should be ClassType

**Comment by Aryant-Tripathi** [comment]
I didn't know about that,  Really sorry @jsotuyod @oowekyala sir, 
Added changes with your suggestions as:  

Support `Boolean` wrapper class in BooleanGetMethodName rule (#5253)

- Updated XPath rule to include both `boolean` primitive and `Boolean` wrapper class:
  (PrimitiveType[@Kind = 'boolean'] or ClassType[pmd-java:typeIs('java.lang.Boolean')])
- Added test cases to ensure that methods returning `Boolean` are also flagged correctly.
- Ensured the rule enforces consistent method naming for both primitive and wrapper types.

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">Compared to main:<br />
This changeset changes 84 violations,<br />
introduces 63 new violations, 0 new errors and 0 new configuration errors,<br />
removes 0 violations, 2 errors and 7 configuration errors.<br />
<a href="https://github.com/pmd/pmd/actions/runs/11276949649?pr=5269">Download full report as build artifact</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by jsotuyod** [comment]
@Aryant-Tripathi thanks, this looks good!

**Review by jsotuyod — APPROVED** [review]

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml [hunk 1]
```diff
@@ -60,4 +70,54 @@ public class Foo {
 }
         ]]></code>
     </test-code>
+
+    <test-code>
+        <description>Bad name</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml:75]:
```suggestion
        <description>Bad name with boxed Boolean (#5253)</description>
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml [hunk 2]
```diff
@@ -60,4 +70,54 @@ public class Foo {
...
+        <expected-problems>1</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public Boolean getEnabled();
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Good name</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml:85]:
```suggestion
        <description>Good name with boxed Boolean (#5253)</description>
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml [hunk 3]
```diff
@@ -34,6 +34,16 @@ public class Foo {
         ]]></code>
     </test-code>
 
+    <test-code>
+        <description>Should not match on multiple parameters by default</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml:38]:
```suggestion
        <description>Should not match for boxed Boolean on multiple parameters by default (#5253)</description>
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml [hunk 4]
```diff
@@ -34,6 +34,16 @@ public class Foo {
...
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public Boolean getEnabled(boolean thisIsNotABean);
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml:42]:
```suggestion
    public Boolean getEnabled(boolean thisIsNotABean) { return false; }
```

I know, the code of the existing test cases won't compile either, but if we create new test cases, we should make sure, we write valid code. It doesn't matter so much for PMD at the moment, as we are much more lenient than a real compiler, but that might not always be the case.

This applies also for the other test cases you added.

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml [hunk 5]
```diff
@@ -60,4 +70,54 @@ public class Foo {
...
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+public class Foo {
+    public Boolean isEnabled();
+}
+        ]]></code>
+    </test-code>
+
+    <test-code>
+        <description>Should not match on methods annotated with @Override</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml:95]:
```suggestion
        <description>Should not match for boxed Boolean on methods annotated with @Override (#5253)</description>
```

File: pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml [hunk 6]
```diff
@@ -60,4 +70,54 @@ public class Foo {
...
+interface Toggleable {
+    Boolean getEnabled(); // NOPMD
+}
+        ]]></code>
+    </test-code>
+
+
+
+    <test-code>
+        <description>Should match on multiple parameters when checkParameterizedMethods = true</description>
```
Comment (adangel) [inline pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/codestyle/xml/BooleanGetMethodName.xml:113]:
```suggestion
        <description>Should match for boxed Boolean on multiple parameters when checkParameterizedMethods = true (#5253)</description>
```

**Review by adangel — COMMENTED** [review]
Thanks for your PR! I've only some minor improvement comments, could you have a look at these?

Also, could you add a new test case, so that we make sure, we don't accidentally match methods that return a custom Boolean type... (we shouldn't, because of `pmd-java:typeIs('java.lang.Boolean')`), but a test case is kind of proof that it works as we expect:

```xml
    <test-code>
        <description>Custom Boolean type (#5253)</description>
        <expected-problems>0</expected-problems>
        <code><![CDATA[
class Boolean { }
public class Foo {
    public Boolean getEnabled() { return null; }
}
        ]]></code>
    </test-code>
```

**Comment by Aryant-Tripathi** [comment]
@adangel sir, I'd like you to please check now and add your suggestions. I hope it resolves issues, still if you find anything missing please suggest changes.

**Review by adangel — APPROVED** [review]
Thanks, looks good now!

**Comment by phansys** [comment]
Thank you @Aryant-Tripathi!

**Pull request merged**